components/blitz/src/ome/services/roi/GeomTool.java

```diff
@@ -354,6 +358,143 @@ public void handle(int x, int y) {
 
     }
 
+    public ShapeStats [] getStatsRestricted(
+            List<Long> shapeIds, 
+            int zForUnattached, int tForUnattached,
+            int[] channels) {
+        if (shapeIds == null || shapeIds.isEmpty()) 
+            throw new ApiUsageException("Provide a non empty list of shape ids.");
+
+        final Session session = factory.getSession();
+        final List<ShapeStats> shapeStats = 
+            new ArrayList<ShapeStats>(shapeIds.size());
+        // we lump together shapes that belong to the same z/t
+        // in order to not have to read the same z/t twice
+        final Map<String, List<ome.model.roi.Shape>> zt_lookup =
+            new HashMap<String, List<ome.model.roi.Shape>>();
+
+       // fetch shapes from db and perform some basic checks
+       List results = 
+           session.createQuery(
+               "select distinct s from Shape s " +
+               "left join fetch s.transform t left join fetch s.roi r " +
+               "join fetch r.image i join fetch i.pixels p " +
+               "where s.id in (:ids)").
+           setParameterList("ids", shapeIds).list();
+       if (results.size() != shapeIds.size()) {
+           throw new ApiUsageException("Given shape id(s) invalid");
+       }
+
+       ome.model.core.Image image = null;
+       ome.model.core.Pixels pixels = null;
+       for (final Object r : results) {
+           final ome.model.roi.Shape shape = (ome.model.roi.Shape) r;
+           final ome.model.roi.Roi roi = shape.getRoi();
+           final ome.model.core.Image img = roi.getImage();
+           
+           // check if all shapes come from the same image
+           if (image == null) {
+               image = img;
+               pixels = image.getPrimaryPixels();
+           } else if (!image.getId().equals(img.getId()))
+               throw new ApiUsageException("All shapes have to be from the same image");
+           // check if z/t unattached fallback values are not out of bounds
+           if (zForUnattached < 0 || zForUnattached >= pixels.getSizeZ() ||
+               tForUnattached < 0 || tForUnattached >= pixels.getSizeT())
+               throw new ApiUsageException(
+                   "Fallback value(s) for unattached z/t shapes are out of bounds");
```

Comment:
should we go for something less aggressive and set the value to ``0 if < 0``
or to ``pixels.getSizeZ/T-1`` if >=pixels.getSizeZ/T``
since we support ``null`` array of channels?

Reply:
@jburel sorry I miss the context, how would a channel null impact here. the z/tForUnattached are meant to be fallback values if a roi is unattached so that we don't iterate over 100 * 60 * planeX * planeY x channels) pixels. the channels come in as a list and have been checked for bounds as well. are you asking if we want to allow unattached (at the cost of potentially doing all the pixel scanning)?

Reply:
``int[] channels`` can be ``null`` and in that case stats for all channels are calculated
We now starts to have images with a large number of channels
i was suggesting that if the value of z is out of bounds then z ==0 or z =max-1 (same for t)
to follow the logic for channels
or we say that "unattached" channels (i.e. null) is not supported

Reply:
I think we need to discuss this. I still see 2 things here, the channel number and the unattached z/t.

If we are concerned about the default behavior for missing channel input (it being filled up with all available channels) and images with a big number of channels we should make that parameter mandatory I believe. Then the caller has to list the channels they need.

As for the fallback of unattached we can default to either 0 or max if the supplied value makes no sense, or anything in between for that matter, if throwing an exception is too radical.